File: src/colav_automaton/controllers/virtual_waypoint.py
"""
Virtual waypoint computation for COLREGs-compliant collision avoidance.

Implements the V1 selection from the paper (Section 4.1, 4.3):
- V1 is chosen as the vertex of the unsafe set with the largest relative
  heading angle toward starboard (or port for rule 17b).
- Both starboard and port candidates are evaluated; the one yielding the
  larger predicted CPA is selected.
"""

from typing import List, Tuple, Optional, Callable
import numpy as np


def _normalize_angle(angle: float) -> float:
    """Normalize angle to [-pi, pi]."""
    return np.arctan2(np.sin(angle), np.cos(angle))


def _predicted_cpa(pos_x, pos_y, v, heading, ox, oy, ovx, ovy):
    """Predict CPA distance if ego heads in the given direction."""
    new_vx = v * np.cos(heading)
    new_vy = v * np.sin(heading)
    dx, dy = ox - pos_x, oy - pos_y
    dvx = ovx - new_vx
    dvy = ovy - new_vy
    dv_sq = dvx ** 2 + dvy ** 2
    if dv_sq < 1e-6:
        return np.hypot(dx, dy)
    t = max(-(dx * dvx + dy * dvy) / dv_sq, 0.0)
    cx = pos_x + new_vx * t
    cy = pos_y + new_vy * t
    return np.hypot(cx - (ox + ovx * t), cy - (oy + ovy * t))


def _min_cpa_for_vertex(pos_x, pos_y, v, vx, vy, obstacles_list):
    """Return minimum predicted CPA across all obstacles if heading toward (vx, vy)."""
    heading = np.arctan2(vy - pos_y, vx - pos_x)
    min_cpa = np.inf
    for ox, oy, ov, o_psi in obstacles_list:
        cpa = _predicted_cpa(pos_x, pos_y, v, heading,
                             ox, oy, ov * np.cos(o_psi), ov * np.sin(o_psi))
        min_cpa = min(min_cpa, cpa)
    return min_cpa


def _point_in_polygon(px: float, py: float, vertices: List[Tuple[float, float]]) -> bool:
    """Check if point (px, py) is inside a polygon using ray casting."""
    if len(vertices) < 3:
        return False

    n = len(vertices)
    inside = False

    j = n - 1
    for i in range(n):
        xi, yi = vertices[i]
        xj, yj = vertices[j]

        if ((yi > py) != (yj > py)) and (px < (xj - xi) * (py - yi) / (yj - yi) + xi):
            inside = not inside
        j = i

    return inside


def _polygon_centroid(vertices: List[Tuple[float, float]]) -> Tuple[float, float]:
    """Compute centroid (geometric center) of a polygon."""
    if not vertices:
        return (0.0, 0.0)

    centroid_x = sum(v[0] for v in vertices) / len(vertices)
    centroid_y = sum(v[1] for v in vertices) / len(vertices)
    return (centroid_x, centroid_y)


def _offset_point_from_centroid(
    point_x: float,
    point_y: float,
    centroid_x: float,
    centroid_y: float,
    offset_distance: float
) -> Tuple[float, float]:
    """Move a point outward from a centroid by a specified distance."""
    outward = np.array([point_x - centroid_x, point_y - centroid_y])
    outward_len = np.linalg.norm(outward)

    if outward_len < 1e-6:
        return (point_x, point_y)

    outward_normalized = outward / outward_len
    new_x = point_x + offset_distance * outward_normalized[0]
    new_y = point_y + offset_distance * outward_normalized[1]

    return (new_x, new_y)


def default_vertex_provider(
    pos_x: float,
    pos_y: float,
    obstacles_list: List[Tuple[float, float, float, float]],
    Cs: float,
    psi: float = 0.0
) -> Optional[List[Tuple[float, float]]]:
    """
    Default vertex provider using circular obstacle approximation.

    Creates 8 vertices around each obstacle at distance Cs.

    Args:
        pos_x, pos_y: Ship position (unused in circular approximation)
        obstacles_list: List of (ox, oy, ov, o_psi) tuples
        Cs: Safety radius around obstacles
        psi: Ship heading (unused in circular approximation)

    Returns:
        List of (vx, vy) vertices or None if no obstacles
    """
    if not obstacles_list:
        return None

    vertices = []
    for ox, oy, _, _ in obstacles_list:
        for i in range(8):
            angle = i * np.pi / 4
            vertices.append((ox + Cs * np.cos(angle), oy + Cs * np.sin(angle)))

    return vertices if vertices else None


def compute_v1(
    pos_x: float,
    pos_y: float,
    psi: float,
    obstacles_list: List[Tuple[float, float, float, float]],
    Cs: float,
    vertex_provider: Callable[
        [float, float, List[Tuple[float, float, float, float]], float, float],
        Optional[List[Tuple[float, float]]]
    ],
    buffer_distance: float = 0.0,
    v: float = 12.0,
) -> Optional[Tuple[float, float]]:
    """
    Compute virtual waypoint V1 from unsafe set vertices (paper Section 4.1).

    Per the paper, V1 is the vertex of the unsafe set for which the relative
    heading angle ρ(V_i, p) is largest.  Starboard vertices (V1, V2 in the
    paper) have negative relative angles; port vertices (V3, V4) have positive.

    This implementation evaluates BOTH the starboard-most and port-most ahead
    vertices, predicts the CPA each would yield, and selects the one with the
    larger CPA.  This extends the paper to handle rule 17b (stand-on vessel
    taking late action to port when give-way vessel fails to act).

    Args:
        pos_x, pos_y: Current ship position
        psi: Current heading (radians)
        obstacles_list: List of (ox, oy, ov, o_psi) tuples
        Cs: Safety distance from obstacles
        vertex_provider: Function to generate unsafe set vertices
        buffer_distance: Optional buffer distance to apply (default 0.0)
        v: Ship velocity for CPA prediction (m/s)

    Returns:
        Tuple (v1_x, v1_y) or None if no valid vertex ahead
    """
    if not obstacles_list:
        return None

    vertices = vertex_provider(pos_x, pos_y, obstacles_list, Cs, psi)
    if not vertices:
        return None

    # Find starboard-most and port-most ahead vertices (paper Section 4.1)
    stbd_vertex = None
    stbd_angle = np.inf  # most negative = most starboard
    port_vertex = None
    port_angle = -np.inf  # most positive = most port

    for vx, vy in vertices:
        angle_to_vertex = np.arctan2(vy - pos_y, vx - pos_x)
        relative_angle = _normalize_angle(angle_to_vertex - psi)

        # Only vertices ahead (within ±90° of heading, paper eq 16 criterion)
        if -np.pi / 2 < relative_angle < np.pi / 2:
            # Starboard: most negative relative angle
            if relative_angle < stbd_angle:
                stbd_angle = relative_angle
                stbd_vertex = (vx, vy)
            # Port: most positive relative angle
            if relative_angle > port_angle:
                port_angle = relative_angle
                port_vertex = (vx, vy)

    # Evaluate CPA for both candidates
    stbd_cpa = (_min_cpa_for_vertex(pos_x, pos_y, v,
                                     stbd_vertex[0], stbd_vertex[1], obstacles_list)
                if stbd_vertex is not None else -1.0)
    port_cpa = (_min_cpa_for_vertex(pos_x, pos_y, v,
                                     port_vertex[0], port_vertex[1], obstacles_list)
                if port_vertex is not None else -1.0)

    if stbd_vertex is None and port_vertex is None:
        return None

    # COLREGs preference: choose starboard unless port gives significantly
    # better CPA (> 10% improvement).  This ensures starboard is the default
    # per Rules 13-15, while allowing port (rule 17b) when it's clearly better.
    if stbd_vertex is not None and (port_vertex is None or port_cpa <= stbd_cpa * 1.1):
        best_vertex = stbd_vertex
    else:
        best_vertex = port_vertex

    if buffer_distance > 0:
        best_vertex = _apply_v1_buffer(
            best_vertex[0], best_vertex[1], vertices, obstacles_list,
            buffer_distance
        )


    return best_vertex


def _apply_v1_buffer(
    v1_x: float,
    v1_y: float,
    vertices: List[Tuple[float, float]],
    obstacles_list: List[Tuple[float, float, float, float]],
    buffer_distance: float,
) -> Tuple[float, float]:
    """
    Apply buffer to V1 by moving it outward from polygon centroid.

    The buffer is rejected (original V1 returned) if:
    - Buffered V1 would be inside the unsafe polygon
    - Buffered V1 would be closer to any obstacle than original V1
    """
    if buffer_distance <= 0 or not vertices or len(vertices) < 3:
        return (v1_x, v1_y)

    centroid_x, centroid_y = _polygon_centroid(vertices)
    buffered_x, buffered_y = _offset_point_from_centroid(
        v1_x, v1_y, centroid_x, centroid_y, buffer_distance
    )

    # Reject if offset failed
    if (buffered_x, buffered_y) == (v1_x, v1_y):
        return (v1_x, v1_y)

    # Reject if buffered V1 would be inside polygon
    if _point_in_polygon(buffered_x, buffered_y, vertices):
        return (v1_x, v1_y)

    # Reject if buffered V1 is closer to any obstacle
    for ox, oy, _, _ in obstacles_list:
        orig_dist = np.hypot(v1_x - ox, v1_y - oy)
        buffered_dist = np.hypot(buffered_x - ox, buffered_y - oy)
        if buffered_dist < orig_dist - 0.1:
            return (v1_x, v1_y)

    return (buffered_x, buffered_y)

File: ros2/src/colav_cpp/src/virtual_waypoint.cpp
#include "colav_cpp/virtual_waypoint.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace colav {

namespace {

constexpr double kPi = M_PI;
const double kInf = std::numeric_limits<double>::infinity();

double normalize_angle(double a) {
  return std::atan2(std::sin(a), std::cos(a));
}

void add_circle(std::vector<Point>& out, double xc, double yc, double r,
                int n = 10) {
  for (int i = 0; i < n; ++i) {
    const double theta = i * (2.0 * kPi / n);
    out.push_back({xc + r * std::cos(theta), yc + r * std::sin(theta)});
  }
}

// 8 vertices per obstacle at v1_Cs — the default_vertex_provider fallback.
std::vector<Point> default_vertices(const std::vector<Obstacle>& obstacles,
                                    double Cs) {
  std::vector<Point> v;
  for (const auto& o : obstacles) {
    for (int i = 0; i < 8; ++i) {
      const double a = i * (kPi / 4.0);
      v.push_back({o[0] + Cs * std::cos(a), o[1] + Cs * std::sin(a)});
    }
  }
  return v;
}

// Predicted CPA if the ship heads toward (vx, vy), mirroring _predicted_cpa /
// _min_cpa_for_vertex.
double predicted_cpa(double px, double py, double v, double heading,
                     double ox, double oy, double ovx, double ovy) {
  const double new_vx = v * std::cos(heading);
  const double new_vy = v * std::sin(heading);
  const double dx = ox - px;
  const double dy = oy - py;
  const double dvx = ovx - new_vx;
  const double dvy = ovy - new_vy;
  const double dv_sq = dvx * dvx + dvy * dvy;
  if (dv_sq < 1e-6) return std::hypot(dx, dy);
  const double t = std::max(-(dx * dvx + dy * dvy) / dv_sq, 0.0);
  const double cx = px + new_vx * t;
  const double cy = py + new_vy * t;
  return std::hypot(cx - (ox + ovx * t), cy - (oy + ovy * t));
}

double min_cpa_for_vertex(double px, double py, double v, double vx, double vy,
                          const std::vector<Obstacle>& obstacles) {
  const double heading = std::atan2(vy - py, vx - px);
  double min_cpa = kInf;
  for (const auto& o : obstacles) {
    const double ovx = o[2] * std::cos(o[3]);
    const double ovy = o[2] * std::sin(o[3]);
    min_cpa = std::min(min_cpa,
                       predicted_cpa(px, py, v, heading, o[0], o[1], ovx, ovy));
  }
  return min_cpa;
}

}  // namespace

namespace {

// One entry of the augmented (swept) obstacle list create_unsafe_set sees.
struct Aug { double x, y, vel, heading; };

}  // namespace

std::vector<Point> swept_unsafe_hull(double px, double py, double psi,
                                     const std::vector<Obstacle>& obstacles,
                                     double v1_Cs, double ship_v,
                                     double dsafe, double max_horizon) {
  if (obstacles.empty()) return {};

  // _compute_swept_obstacles: original obstacle (its velocity) plus
  // zero-velocity trajectory samples up to min(tcpa, max_horizon).
  std::vector<Aug> aug;
  for (const auto& o : obstacles) {
    const double ox = o[0], oy = o[1], ov = o[2], opsi = o[3];
    aug.push_back({ox, oy, ov, opsi});
    const Cpa c = calc_cpa(px, py, psi, ship_v, ox, oy, ov, opsi);
    if (ov > 0.1 && c.tcpa > 0.0) {
      const double sweep = std::min(c.tcpa, max_horizon);
      int n = static_cast<int>(sweep / 10.0);
      n = std::max(3, std::min(20, n));
      for (int i = 1; i <= n; ++i) {
        const double dt = sweep * i / n;
        aug.push_back({ox + ov * std::cos(opsi) * dt,
                       oy + ov * std::sin(opsi) * dt, 0.0, opsi});
      }
    }
  }

  // create_unsafe_set on the augmented list with dsf = dsafe. Unlike the G11
  // path (large distance_safety -> all obstacles), here the index filters
  // genuinely prune distant samples. Agent and obstacle safety radius are
  // both v1_Cs, so every distance test subtracts 2*v1_Cs.
  const int k = static_cast<int>(aug.size());
  std::vector<Cpa> m(k);
  std::vector<bool> i1(k), i3(k);
  for (int i = 0; i < k; ++i) {
    m[i] = calc_cpa(px, py, psi, ship_v, aug[i].x, aug[i].y, aug[i].vel,
                    aug[i].heading);
    const double d_agent = std::hypot(aug[i].x - px, aug[i].y - py);
    i1[i] = (d_agent - 2.0 * v1_Cs) <= dsafe;
    i3[i] = (m[i].dcpa <= dsafe) && (m[i].tcpa <= 15.0);
  }
  // I2: obstacle in I1 with another obstacle within dsafe.
  std::vector<bool> i2(k, false);
  for (int i = 0; i < k; ++i) {
    if (!i1[i]) continue;
    for (int j = 0; j < k; ++j) {
      if (j == i) continue;
      const double d = std::hypot(aug[i].x - aug[j].x, aug[i].y - aug[j].y);
      if ((d - 2.0 * v1_Cs) <= dsafe) { i2[i] = true; break; }
    }
  }

  std::vector<Point> verts;
  for (int i = 0; i < k; ++i) {
    if (!(i1[i] || i2[i] || i3[i])) continue;  // uIoI membership
    add_circle(verts, aug[i].x, aug[i].y, v1_Cs);
    if (m[i].tcpa > 0.0 && !std::isnan(m[i].tcpa)) {
      add_circle(verts, aug[i].x + aug[i].vel * std::cos(aug[i].heading) * m[i].tcpa,
                 aug[i].y + aug[i].vel * std::sin(aug[i].heading) * m[i].tcpa, v1_Cs);
    }
  }
  return convex_hull(std::move(verts));
}

std::optional<Point> compute_v1(double px, double py, double psi,
                                const std::vector<Obstacle>& obstacles,
                                double v1_Cs, double ship_v, double dsafe,
                                double max_horizon, double buffer_distance) {
  if (obstacles.empty()) return std::nullopt;

  std::vector<Point> verts = swept_unsafe_hull(px, py, psi, obstacles, v1_Cs,
                                               ship_v, dsafe, max_horizon);
  if (verts.empty()) verts = default_vertices(obstacles, v1_Cs);
  if (verts.empty()) return std::nullopt;

  // Starboard-most (most negative) and port-most (most positive) vertex that
  // lies ahead (within +/- pi/2 of heading).
  const Point* stbd = nullptr;
  double stbd_angle = kInf;
  const Point* port = nullptr;
  double port_angle = -kInf;
  for (const auto& vtx : verts) {
    const double rel = normalize_angle(std::atan2(vtx.y - py, vtx.x - px) - psi);
    if (-kPi / 2.0 < rel && rel < kPi / 2.0) {
      if (rel < stbd_angle) { stbd_angle = rel; stbd = &vtx; }
      if (rel > port_angle) { port_angle = rel; port = &vtx; }
    }
  }
  if (stbd == nullptr && port == nullptr) return std::nullopt;

  const double stbd_cpa =
      stbd ? min_cpa_for_vertex(px, py, ship_v, stbd->x, stbd->y, obstacles)
           : -1.0;
  const double port_cpa =
      port ? min_cpa_for_vertex(px, py, ship_v, port->x, port->y, obstacles)
           : -1.0;

  // COLREGs starboard preference: port only when clearly (>10%) better.
  Point best;
  if (stbd != nullptr && (port == nullptr || port_cpa <= stbd_cpa * 1.1)) {
    best = *stbd;
  } else {
    best = *port;
  }

  (void)buffer_distance;  // buffer (>0) not exercised here; CommonOcean uses 0
  return best;
}

}  // namespace colav
